### DISCUSSION TITLE
Anyone using bucket4j with Jakarta EE servers, e.g. Wildfly?

### DISCUSSION THREAD

**Discussion by helge79** [vladimir-bukhtoyarov/bucket4j#516] [Q&A] [answered]
Hi,

We have a Jakarta EE application for which we intend to use bucket4j - across a clustered setup. At the customer's request, our application server has to be Wildfly 32, which uses Infinispan under the hood. So setup of cache containers and caches happens there. Our cache is currently setup as:

```xml
<subsystem xmlns="urn:jboss:domain:infinispan:14.0">
[..]
    <cache-container name="foo" default-cache="bar">
        <transport lock-timeout="3000"/>
        <replicated-cache name="bar"/>
    </cache-container>
</subsystem>
```

Cache is then injected in EE manner as `@Resource`:

```java
import org.infinispan.Cache;
[...]
@Resource(lookup = "java:jboss/infinispan/cache/foo/bar")
private Cache<String, byte[]> cache;	
```

So I have no clue how to access the `ReadWriteMap` from here which `bucket4j-infinispan-all` would work with. Neither do I think that Hot Rod is applicable here since Infinispan is embedded in Wildfly. I am also not able to use `bucket4j-jcache` since JCache still is not part of EE 10 (hooray 🙄). So I came up with this which is a mixture of implementations of `bucket4j-caffeine` and `bucket4j-jcache`:

```java
public class InfinispanCacheProxyManager<K> extends AbstractProxyManager<K> {

  private final Cache<K, byte[]> cache; 
  [...]

  @Override
  protected <T> CommandResult<T> execute(K key, Request<T> request) {
    Long clientSideTime = request.getClientSideTime();
    long timeNanos = clientSideTime != null ? clientSideTime : System.currentTimeMillis() * 1_000_000;

    byte[] serializedState = cache.get(key);
    RemoteBucketState oldState = serializedState == null ? null : InternalSerializationHelper.deserializeState(serializedState);

    MutableBucketEntry entryWrapper = new MutableBucketEntry(oldState == null ? null : oldState.copy());
    CommandResult<T> result = request.getCommand().execute(entryWrapper, timeNanos);
    if (entryWrapper.exists()) {
      serializedState = InternalSerializationHelper.serializeState(entryWrapper.get(), request.getBackwardCompatibilityVersion());
      cache.put(key, serializedState);
    }
    return result;
  } 

  [...]
}
```

It works, but it doesn't feel like it's the 100% right solution. Apart from being not atomic, which would be kind of acceptable for our use case, could this implmentation cause any other issues I am not aware of? I also tried to use Infinispan's `Cache.compute()` which takes a `org.infinispan.util.function.SerializableBiFunction` and executes it in an atomic way but couldn't make it to do work with Wildfly. I experienced several `ClassNotFoundException` here although I tried several manifest dependencies. No idea why, the documentation on the Wildfly and Infinispan website is unfortunately very scarce on the whole "use a cache in your EE app" subject.

So final question, does anyone use `bucket4j` in a Jakarta EE application server, preferably Wildfly, and can share his implementation and/or point me in the correct direction to achieve atomic operation?

Thank you for any answers or hints.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @helge79 

[This unit test](https://github.com/bucket4j/bucket4j/blob/master/bucket4j-infinispan-all/bucket4j-infinispan/src/test/java/io/github/bucket4j/grid/infinispan/InfinispanTest.java#L80) can be helpful to figure out the way of getting `org.infinispan.functional.impl.ReadWriteMapImpl` from `org.infinispan.Cache`. Bellow just a copy&past of required code fragment:
```java
private static ReadWriteMap<String, byte[]> toMap(Cache<String, byte[]> cache) {
        FunctionalMapImpl<String, byte[]> functionalMap = FunctionalMapImpl.create(cache.getAdvancedCache());
        return ReadWriteMapImpl.create(functionalMap);
}
```

Do not forget to configure serialization https://bucket4j.com/8.14.0/toc.html#special-notes-for-infinispan-10-0

**Reply by helge79** [reply]
Hi @vladimir-bukhtoyarov,

ah, it was that easy to access the `FunctionalMapImpl`? Damn. Thank you.

For anyone who stumbles over a similar issue, some more notes. I again experienced `ClassNotFoundException` so I needed to add some more module dependencies, namely `org.infinispan.core` and `org.infinispan.common`. This is the excerpt from my EJB's `pom.xml` for my EJB in an EAR (hence the "export") which makes `bucket4j-infispan` work perfectly.

```xml
<archive>
  <manifestEntries>
    <Dependencies>org.infinispan export, org.infinispan.core export, org.infinispan.commons export</Dependencies>
  </manifestEntries>
</archive>
```

Another thing to notice as @vladimir-bukhtoyarov pointed out. The default marshaller in Wildfly for a cache container is LEGACY [as documented here](https://docs.wildfly.org/32/wildscribe/subsystem/infinispan/cache-container/index.html). If you have changed this to PROTOBUF, you should either add a separate cache container or set the marshaller accordingly.
